File: 프로그래머스/3/43162. 네트워크/네트워크.java
class Solution {
    int[][] computers;
    boolean[] visited;
    
    public int solution(int n, int[][] computers) {
        this.computers = computers;
        this.visited = new boolean[computers.length];
        
        int result = 0;
        for(int i = 0; i < visited.length; i++) {
            if(!visited[i]) {
                dfs(i);
                result++;
            }
        }
        
        return result;
    }
    
    private void dfs(int node) {
        visited[node] = true;
        
        for(int i = 0; i < computers[node].length; i++) {
            if(!visited[i] && computers[node][i] == 1) {
                dfs(i);
            }
        }
    }
}

File: 프로그래머스/3/43162. 네트워크/네트워크.cpp
#include <string>
#include <vector>
#include <queue>
#define MAX 200

using namespace std;

vector<vector<int>> computers_;
bool visited[MAX];

void bfs(int i) {
    queue<int> q;
    q.push(i);
    visited[i] = true;
    
    while (!q.empty()) {
        int now = q.front();
        q.pop();
        
        for (int j = 0; j < computers_[now].size(); j++) {
            if (!visited[j] && computers_[now][j] == 1) {
                visited[j] = true;
                q.push(j);
            }
        }
    }
}

int solution(int n, vector<vector<int>> computers) {
    int answer = 0;
    computers_ = computers;
    
    for (int i = 0; i < n; i++) {
        if (!visited[i]) {
            bfs(i);
            answer++;
        }
    }
    
    return answer;
}